### DISCUSSION TITLE
How do I export an environment variable with wait?

### DISCUSSION THREAD

**Discussion by xenoterracide** [z-shell/zi#288] [Q&A]
I have this, but `$JAVA_HOME` isn't being exported to my shell

```zsh
zi ice lucid wait'[[ -f ~/.asdf/plugins/java/set-java-home.zsh ]]' has'asdf' \
  atload'. ~/.asdf/plugins/java/set-java-home.zsh' for z-shell/null
```

I imagine it has something to do with subshells... or something.

**Comment by ss-o** [comment]
Hi @xenoterracide :wave:

The issue is in syntax as `zi ice ...` and the `for` are meant to be used separately.

I am unfamiliar with `asdf`, but here is my suggestion/example of loading the `set-java-home.zsh` with Zi:

```shell
zi lucid id-as'z-shell/init-java' \
  wait'[[ -n ${ZLAST_COMMANDS[(r)java*]} ]]' has'asdf' \
  atload'. ~/.asdf/plugins/java/set-java-home.zsh' for z-shell/0
```

- The `id-as'z-shell/init-java'` is used to set a unique ID not confuse it with the `z-shell/0` repository itself as it may prevent loading the plugin because of duplicate ID.

- The `[[ -n ${ZLAST_COMMANDS[(r)java*]} ]]` is used to load the plugin/snippet only when typing `java` instead of loading it every time the shell starts.

> Reference: https://wiki.zshell.dev/docs/guides/syntax/standard#wait

**Reply by xenoterracide** [reply]
but what if I need it for more than just running java? I mean, I almost never run java directly. I usually am running things like gradle or maven

**Reply by xenoterracide** [reply]
I tried this instead

```zsh
zi wait atload=". /opt/asdf-vm/asdf.sh" for z-shell/null
zi lucid id-as'z-shell/init-java' \
  wait'1' has'asdf' \
  atload'. ~/.asdf/plugins/java/set-java-home.zsh' for z-shell/0
```
```
❯ echo $JAVA_HOME

❯ cat ~/.zshrc
❯ . ~/.asdf/plugins/java/set-java-home.zsh
❯ echo $JAVA_HOME
/home/xeno/.asdf/installs/java/temurin-17.0.8+7
```

**Reply by ss-o** [reply]
If you could provide all the information and steps on how and when you would like to export the variable (initiate the script), then maybe I would have an idea to propose :speak_no_evil: 

Zi will initiate whenever the condition in the `wait' ... '` ice is true, so it can be location, time, etc. For example you `cd` to your work environment or some specific directory/path :sparkles:

**Reply by xenoterracide** [reply]
if what I have given isn't enough. I would suggest maybe reading https://asdf-vm.com/guide/getting-started.html and https://github.com/halcyon/asdf-java and then if you want you could set up my entire local environment with yadm (no promises it works without modification for you, esp bootstrap) https://yadm.io/docs/install all of my configuration is here https://github.com/xenoterracide/dotfiles although you'd probably want to start reading here https://github.com/xenoterracide/dotfiles/tree/master/.config/yadm/alt

I don't know what you're asking me for at this point. 

In order for `mvn` to successfully run $JAVA_HOME has to be set before you try.

**Comment by ss-o** [comment]
> 
> In order for `mvn` to successfully run $JAVA_HOME has to be set before you try.

Such key details should be mentioned when describing an issue :point_up: The discussion question was "How do I export an environment variable with wait?" Missing context got me confused.

However, most importantly the issue seems to be solved :dart: :1st_place_medal: :upside_down_face:

**Reply by caleb-cushing0-kr** [reply]
not entirely, it still doesn't seem to set that env var the first time the prompt loads, only after I have a second prompt.
